lib/puppetserver/ca/action/generate.rb

```diff
@@ -202,13 +202,22 @@ def generate_certs(certnames, alt_names, settings, digest)
             key, csr = generate_key_csr(certname, settings, digest, current_alt_names)
             next false unless csr
             next false unless ca.submit_certificate_request(certname, csr)
-            next false unless ca.sign_certs([certname])
+
+            # Check if the CA autosigned the cert
             if result = ca.get_certificate(certname)
+              @logger.inform "Certificate for #{certname} was autosigned."
               next false unless save_file(result.body, certname, settings[:certdir], "Certificate")
               next false unless save_keys(certname, settings, key)
               true
             else
-              false
+              next false unless ca.sign_certs([certname])
+              if result = ca.get_certificate(certname)
+                next false unless save_file(result.body, certname, settings[:certdir], "Certificate")
+                next false unless save_keys(certname, settings, key)
```

Comment:
Did we want to save keys after a successful cert request?

Reply:
I think that's what this does... are you saying we should save them earlier? (e.g. after CSR submission)

Reply:
yeah, sorry, do we want to save the keys after a successful cert request, regardless of the signing outcome?